### PR TITLE
refactor Qwen3-Next with a new RadixLinearAttention

### PR DESCRIPTION
**Pull request by zminglei**
## Motivation
Refactor: Abstract linear attention backend calls through `RadixLinearAttention`

## Modifications

- `qwen3_next.py`: Replaced direct forward_batch.attn_backend.forward() calls with RadixLinearAttention interface

- `hybrid_linear_attn_backend.py`: Backend now extracts layer parameters (conv_weights, A_log, dt_bias, bias, activation, dimensions) from the layer object instead of kwargs

Init-time params stored on layer: `conv_weights, bias, A_log, dt_bias, activation, head dimensions`
Runtime tensors passed via kwargs: `mixed_qkv, a, b, z`

## Accuracy Tests

```
python3 -m sglang.launch_server --model-path /shared/public/elr-models/Qwen/Qwen3-Next-80B-A3B-Instruct/9c7f2fbe84465e40164a94cc16cd30b6999b0cc7/ --tp 4

python benchmark/gsm8k/bench_sglang.py --data-path /shared/public/data/gsm8k/test.jsonl
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:20<00:00,  9.57it/s]
Accuracy: 0.960
Invalid: 0.000
Latency: 21.029 s
Output throughput: 1522.310 token/s
```

With piecewise cuda graph:
```
python3 -m sglang.launch_server --model-path /shared/public/elr-models/Qwen/Qwen3-Next-80B-A3B-Instruct/9c7f2fbe84465e40164a94cc16cd30b6999b0cc7/ --tp 4 --enable-piecewise-cuda-graph

python benchmark/gsm8k/bench_sglang.py --data-path /shared/public/data/gsm8k/test.jsonl
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:14<00:00, 13.86it/s]
Accuracy: 0.955
Invalid: 0.000
Latency: 14.492 s
Output throughput: 2261.951 token/s
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 1]
```diff
@@ -864,20 +864,23 @@ def forward_decode(
...
-        attn_tp_size = kwargs["attention_tp_size"]
-        head_k_dim = kwargs["head_k_dim"]
-        head_v_dim = kwargs["head_v_dim"]
         a = kwargs["a"]
         b = kwargs["b"]
-        A_log = kwargs["A_log"]
-        dt_bias = kwargs["dt_bias"]
-        layer_id = kwargs["layer_id"]
+
+        if layer is not None:
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: is it possible that layer is none?

Comment (zminglei) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: No, layer should always be passed as a non-None value from the RadixLinearAttention, I've removed the check here to avoid confusion.

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 2]
```diff
@@ -831,25 +832,27 @@ def forward_decode(
         q: torch.Tensor,
         k: torch.Tensor,
         v: torch.Tensor,
-        layer: RadixAttention,
+        layer: RadixLinearAttention,
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:832]: change this interface to be

```
    def forward_decode(
        self,
        layer: RadixLinearAttention,
        forward_batch: ForwardBatch,
        mixed_qkv: torch.Tensor,
        a: torch.Tensor,
        b: torch.Tensor,
    )
```

Comment (hebiao064) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:832]: the goal is to remove all kwargs

Comment (hebiao064) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:832]: for the hybrid linear attention backend, we can do 
```
 def forward(
        self,
        layer: RadixLinearAttention,
        forward_batch: ForwardBatch,
        save_kv_cache: bool = True,
        q: Optional[torch.Tensor] = None,  # For full attention
        k: Optional[torch.Tensor] = None,  # For full attention
        v: Optional[torch.Tensor] = None,  # For full attention
        mixed_qkv: Optional[torch.Tensor] = None,  # For Linear Attention
        a: Optional[torch.Tensor] = None,  # For Linear Attention
        b: Optional[torch.Tensor] = None,  # For Linear Attention
        **kwargs,
    )
```

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 3]
```diff
@@ -909,26 +912,30 @@ def forward_extend(
...
-        head_v_dim = kwargs["head_v_dim"]
         a = kwargs["a"]
         b = kwargs["b"]
-        A_log = kwargs["A_log"]
-        dt_bias = kwargs["dt_bias"]
-        layer_id = kwargs["layer_id"]
-        seq_len = kwargs["seq_len"]
+
+        # seq_len can be derived from mixed_qkv or passed explicitly
+        seq_len = kwargs.get("seq_len", mixed_qkv.shape[0])
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: we don't need to pass it from outside, we can just do `seq_len = mixed_qkv.shape[0]`

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 4]
```diff
@@ -909,26 +912,30 @@ def forward_extend(
...
+        bias = layer.bias
+        activation = layer.activation
+        key_dim = layer.key_dim
+        value_dim = layer.value_dim
+        attn_tp_size = layer.attention_tp_size
+        head_k_dim = layer.head_k_dim
+        head_v_dim = layer.head_v_dim
+        A_log = layer.A_log
+        dt_bias = layer.dt_bias
+        layer_id = layer.layer_id
```
Comment (hebiao064) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: some of the variables are only being used once, we don't need to reassign to a new value, for example: `A_log`

File: python/sglang/srt/models/qwen3_next.py [hunk 5]
```diff
@@ -305,6 +304,25 @@ def __init__(
             prefix=add_prefix("out_proj", prefix),
         )
 
+        # Create the 2D view of conv weights once at init
+        conv_weights = self.conv1d.weight.view(
```
Comment (hebiao064) [inline python/sglang/srt/models/qwen3_next.py]: do we need this?

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 2 above)
Comment (zminglei) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:832]: updated it, ptal

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 3 above)
Comment (zminglei) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: updated

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 4 above)
Comment (zminglei) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: updated

File: python/sglang/srt/models/qwen3_next.py (hunk 5 above)
Comment (zminglei) [reply python/sglang/srt/models/qwen3_next.py]: Yes, it's needed inside the GDN attention, I replaced it with a squeeze op.

**Review by hebiao064 — APPROVED** [review]

**Review by yizhang2077 — APPROVED** [review]

**Comment by yizhang2077** [comment]
Could we also move other hybrid models like kimi linear into `RadixLinearAttention` later?

**Review by hebiao064 — APPROVED** [review]

File: python/sglang/srt/layers/radix_linear_attention.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+from typing import TYPE_CHECKING, Optional
+
+import torch
+from torch import nn
+
+if TYPE_CHECKING:
+    from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+
+
+class RadixLinearAttention(nn.Module):
```
Comment (ispobock) [inline python/sglang/srt/layers/radix_linear_attention.py]: Can it also be applied to other linear models like kimi-linear, nemotron?

**Review by ispobock — APPROVED** [review]

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 7]
```diff
@@ -828,30 +829,23 @@ def __init__(self, model_runner: ModelRunner):
...
-
-        layer_cache = self.req_to_token_pool.mamba2_layer_cache(layer_id)
+        conv_weights = layer.conv_weights
+        bias = layer.bias
+        activation = layer.activation
+        key_dim = layer.key_dim
+        value_dim = layer.value_dim
+        attn_tp_size = layer.attention_tp_size
+        head_k_dim = layer.head_k_dim
+        head_v_dim = layer.head_v_dim
```
Comment (yuan-luo) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:846]: Suggest also add:
```
A_log = layer.A_log
dt_bias = layer.dt_bias
```
to make the parameter list consistent.

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 8]
```diff
@@ -1024,7 +1012,7 @@ def forward_extend(
         key = key.view(1, actual_seq_len, num_heads, head_k_dim)
         value = value.view(1, actual_seq_len, num_value_heads, head_v_dim)
 
-        g, beta = fused_gdn_gating(A_log, a, b, dt_bias)
+        g, beta = fused_gdn_gating(layer.A_log, a, b, layer.dt_bias)
```
Comment (yuan-luo) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:1015]: ditto, in that way, we can avoid to use layer.A_log style here and there.

File: python/sglang/srt/models/qwen3_next.py [hunk 9]
```diff
@@ -419,41 +432,12 @@ def _forward(
...
-            "head_v_dim": self.head_v_dim,
-            "a": a,
-            "b": b,
-            "A_log": self.A_log,
-            "dt_bias": self.dt_bias,
-            "layer_id": self.layer_id,
-            "seq_len": seq_len,
-            "num_k_heads": self.num_k_heads,
-            "num_v_heads": self.num_v_heads,
-            "z": z,
```
Comment (yuan-luo) [inline python/sglang/srt/models/qwen3_next.py:447]: @zminglei I realized here `z` was not passed to AttentionBackend. In Qwen3-Next model, `z`'s calculation rely on hidden_states, it is neither passed by kwargs in forward(kwargs is gone) nor can be deduced in RadixLinearAttention initialization as there's no hidden_states. So this value is lost. 
When refactoring Kimi-Linear to reuse this RadixLinearAttention arch I also encountered this circumstance. There are more parameters depend on hidden_states, but there's no way to pass it as kwargs is wiped off. I think this is an issue in this architecture. The kwargs seems to be inevitable. The parameters which can be calculated can be removed from kwargs and put them into LinearAttention initialization phase.
cc: @ispobock @yizhang2077 @hebiao064 

After double confirming the code, GDN and KDA can use the same parameters:
z: norm_gate
a: forget_gate
b: beta
The z don't need to pass into AttentionBackend. z's shape can be deduced from other parameters.

This code refactor is solid.

**Review by yuan-luo — APPROVED** [review]

File: python/sglang/srt/models/qwen3_next.py (hunk 9 above)
Comment (zminglei) [reply python/sglang/srt/models/qwen3_next.py:447]: Yes, the z is not used in the attention_backend, only its shape was used, and the shape could be deduced from other parameters like `mixed_qkv`, so we're not passing z in forward here.

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 7 above)
Comment (zminglei) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:846]: For the variables which are only used once, we didn't reassign them to a new value/variable. See [comment](https://github.com/sgl-project/sglang/pull/17373#discussion_r2709841444).

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 8 above)
Comment (zminglei) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:1015]: same as above.
